### PR TITLE
feat: make invalid piggyback cause unchallenged exit event when it's close to being finalized

### PR DESCRIPTION
**Pull request by pgebal**
Closes https://github.com/omisego/security-issues/issues/33
## Overview
According to: https://github.com/omisego/elixir-omg/blob/master/docs/exit_validation.md user should be warned about invalid exit that is close to finalization. We were doing that only for standard exits. This PR introduces 2 new types of events: `UnchallengedPiggyback` and `UnchallengedNonCanonicalIFE` that warn about invalid piggybacks and unchallenged non canonical ifes that are close to being finalized.

## Changes
Adds two new type of events: `unchallenged_piggyback` and `unchallenged_non_canonical_ife` that are returned in calls to `/status.get`in case there is a unchallenged piggyback or a non-canonical ife that is close to being finalized.

### CODE REVIEW

**Comment by InoMurko** [comment]
I don't think I'm suitable to review this so I'm assigning @achiurizo

File: apps/omg_watcher/test/omg_watcher/integration/in_flight_exit_test.exs [hunk 1]
```diff
@@ -257,7 +257,82 @@ defmodule OMG.Watcher.Integration.InFlightExitTest do
            end) =~ "Invalid in-flight exit finalization"
 
     assert %{"in_flight_exits" => [_], "byzantine_events" => byzantine_events} = WatcherHelper.success?("/status.get")
-    assert [%{"event" => "invalid_piggyback"}] = Enum.filter(byzantine_events, &(&1["event"] != "piggyback_available"))
+
+    assert [%{"event" => "unchallenged_piggyback"}, %{"event" => "invalid_piggyback"}] =
+             Enum.filter(byzantine_events, &(&1["event"] != "piggyback_available"))
+  end
+
+  @tag fixtures: [:in_beam_watcher, :alice, :bob, :mix_based_child_chain, :token, :alice_deposits]
```
Comment (InoMurko) [inline apps/omg_watcher/test/omg_watcher/integration/in_flight_exit_test.exs:265]: we're stepping away from using fixtures. find another way to test this scenarios

File: apps/omg_watcher/test/omg_watcher/integration/in_flight_exit_test.exs [hunk 2]
```diff
@@ -257,7 +257,82 @@ defmodule OMG.Watcher.Integration.InFlightExitTest do
...
+    exit_processor_sla_margin = Application.fetch_env!(:omg_watcher, :exit_processor_sla_margin)
+    DevHelper.wait_for_root_chain_block(eth_height + exit_processor_sla_margin, @timeout)
+
+    IntegrationTest.wait_for_byzantine_event(
+      %OMG.Watcher.Event.UnchallengedPiggyback{}.name,
+      @timeout
+    )
+  end
+
+  @tag fixtures: [:in_beam_watcher, :alice, :bob, :mix_based_child_chain, :token, :alice_deposits]
```
Comment (InoMurko) [inline apps/omg_watcher/test/omg_watcher/integration/in_flight_exit_test.exs:296]: we're stepping away from using fixtures. find another way to test this scenarios

Comment (pgebal) [reply apps/omg_watcher/test/omg_watcher/integration/in_flight_exit_test.exs:296]: I'll try to code it in a cabbage test.
@InoMurko it's a pity I didn't know about that before coding them. Let's have a list ofdon't, things we're stepping away from, like: fixtures, single pipes, etc, so anyone can read it before contributing. I think you are best informed on this stuff. Could you prepare that list?

**Comment by thec00n** [comment]
Can we also update https://github.com/omisego/elixir-omg/blob/master/docs/exit_validation.md#unchallenged_exit-condition to reflect the changes?

File: apps/omg_watcher/test/omg_watcher/integration/in_flight_exit_test.exs (hunk 2 above)
Comment (pgebal) [reply apps/omg_watcher/test/omg_watcher/integration/in_flight_exit_test.exs]: I decided to rely just on unit tests and not code a separate e2e scenario. Those events are already emitted in existing integration and e2e tests. Creating a separate cabbage test just for that would be a lot of hassle.

File: apps/omg_watcher/test/omg_watcher/exit_processor/canonicity_test.exs [hunk 3]
```diff
@@ -323,6 +323,24 @@ defmodule OMG.Watcher.ExitProcessor.CanonicityTest do
       assert {:ok, %{}} = request |> Core.get_competitor_for_ife(processor, txbytes)
     end
 
+    test "detects that invalid piggyback becomes unchallenged exit when sla period passes",
```
Comment (thec00n) [inline apps/omg_watcher/test/omg_watcher/exit_processor/canonicity_test.exs:326]: The description should be around detecting UnchallengedNonCanonicalIFE?

File: apps/omg_watcher/lib/omg_watcher/exit_processor/piggyback.ex [hunk 4]
```diff
@@ -86,39 +86,66 @@ defmodule OMG.Watcher.ExitProcessor.Piggyback do
...
 
-  @spec get_invalid_piggybacks_events(Core.t(), KnownTx.known_txs_by_input_t()) ::
-          list(Event.InvalidPiggyback.t())
-  def get_invalid_piggybacks_events(%Core{in_flight_exits: ifes}, known_txs_by_input) do
-    ifes
-    |> Map.values()
-    |> all_invalid_piggybacks_by_ife(known_txs_by_input)
-    |> group_by_txbytes()
-    |> materials_to_events()
+  @spec get_invalid_piggybacks_events(Core.t(), KnownTx.known_txs_by_input_t(), pos_integer()) ::
```
Comment (pnowosie) [inline apps/omg_watcher/lib/omg_watcher/exit_processor/piggyback.ex:89]: Can we add @doc here. 🙏

**Pull request merged**